src/main/java/org/opensearch/plugin/insights/core/listener/QueryInsightsListener.java

```diff
@@ -91,26 +91,27 @@ public QueryInsightsListener(final ClusterService clusterService, final QueryIns
      * and query insights services.
      *
      * @param metricType {@link MetricType}
-     * @param enabled    boolean
+     * @param isCurrentMetricEnabled boolean
      */
-    public void setEnableTopQueries(final MetricType metricType, final boolean enabled) {
-        boolean isAllMetricsDisabled = !queryInsightsService.isEnabled();
-        this.queryInsightsService.enableCollection(metricType, enabled);
-        if (!enabled) {
-            // disable QueryInsightsListener only if all metrics collections are disabled now.
-            if (!queryInsightsService.isEnabled()) {
+    public void setEnableTopQueries(final MetricType metricType, final boolean isCurrentMetricEnabled) {
+        boolean isTopNFeatureDisabled = !queryInsightsService.isTopNFeatureEnabled();
+        this.queryInsightsService.enableCollection(metricType, isCurrentMetricEnabled);
+
+        if (!isCurrentMetricEnabled) {
+            // disable QueryInsightsListener only if all metrics collections are disabled now
+            // and search query metrics is disabled.
+            if (isTopNFeatureDisabled) {
```

Comment:
should this be `!queryInsightsService.isTopNFeatureEnabled()` ? 
`isTopNFeatureDisabled` is the flag we get before changing a flag in `this.queryInsightsService.enableCollection`. At this time the flag as already be changed.

Reply:
Added some unit tests to make sure we test this logic and refactored the code to make it clearer